### ISSUE TITLE
Refactor registry configuration file

### ISSUE THREAD

**Issue by miminar** [openshift/origin#13568]
Currently, our custom features are configured using the [middleware section](https://docs.docker.com/registry/configuration/#middleware) of the upstream configuration file. It looks like this:

```yml
middleware:
  registry:
    - name: openshift
  repository:
    - name: openshift
      options:
        acceptschema2: false
        pullthrough: true
        mirrorpullthrough: true
        enforcequota: false
        projectcachettl: 1m
        blobrepositorycachettl: 10m
  storage:
    - name: openshift
```

It's rather confusing and hard to explain. There are three different sub-sections, each requiring a map with at least `name: openshift`. We used to panic if not present. Nowadays we tolerate the missing sections and [create them on registry's startup if missing](https://github.com/openshift/origin/blob/9a98905413981e00bc84784c7654ef149e55d49a/pkg/cmd/dockerregistry/dockerregistry.go#L301). If we extend this more, it will become even more complex.

PR #12711 adds a new top-level section to the configuration file which will probably look like this:

```yml
openshift:
  version: 0.1
  metrics:
    enabled: false
    secret: TopSecretToken
```

In a short-term, this will make the configuration further complex. There's a light at the end of the tunnel though.

This issue proposes a follow-up PR that would make the configuration much more simpler. It proposes to move all the middleware settings in to the top-level section, e.g.:

```yml
openshift:
  version: 0.2
  metrics:
    enabled: false
    secret: TopSecretToken
  cache:
    blobrepositoryttl: 1m
    quotattl: 1m    # new name for projectcache ttl
  quota:
    enforce: false
  pullthrough:
    disabled: false
    mirrorblobs: true
  compability:
    acceptschema2: true
```

The final layout layout may look differently. The main thing is the shift from the `middleware` section to the `openshift` one. By this move we hope to make integrated registry config great again.

We will still maintain backwards compatibility with the middleware section for some time. We will continue to support `REGISTRY_MIDDLEWARE_REPOSITORY_*` environment variables as well. However, they will be overriden by `REGISTRY_OPENSHIFT_*` variables if set. These can be parsed using [upstream generic configuration parser](https://github.com/docker/distribution/blob/0700fa570d7bcc1b3e46ee127c4489fd25f4daa3/configuration/parser.go#L84) (thank you Docker team!).

**Comment by miminar** [comment]
@legionus @dmage Feel free to review, suggest improvements or even self-assign.
@mfojtik FYI

**Comment by mfojtik** [comment]
idea sounds good to me

**Comment by bparees** [comment]
@miminar @legionus this was fixed by https://github.com/openshift/image-registry/pull/14 right?

**Comment by miminar** [comment]
> this was fixed by openshift/image-registry#14 right?

That's right, closing.
